src/components/connectAdvanced.js

```diff
@@ -118,146 +82,130 @@ export default function connectAdvanced(
-        // replacing it with a no-op on unmount. This can probably be avoided if Subscription's
-        // listeners logic is changed to not call listeners that have been unsubscribed in the
-        // middle of the notification loop.
-        this.notifyNestedSubs = this.subscription.notifyNestedSubs.bind(this.subscription)
-      }
+      render() {
+        if(this.state.error) {
+          throw this.state.error;
+        }
 
-      onStateChange() {
-        this.runUpdater(this.notifyNestedSubs)
+        return <WrappedComponent {...this.state.childProps} />
       }
+    }
 
-      isSubscribed() {
-        return Boolean(this.subscription) && this.subscription.isSubscribed()
-      }
+    class Connect extends OuterBaseComponent {
+      constructor(props) {
+        super(props)
 
+        this.renderInner = this.renderInner.bind(this);
+      }
+/*
       addExtraProps(props) {
-        if (!withRef && !renderCountProp && !(this.propsMode && this.subscription)) return props
+          if (!withRef && !renderCountProp) return props;
+
         // make a shallow copy so that fields added don't leak to the original selector.
         // this is especially important for 'ref' since that's a reference back to the component
         // instance. a singleton memoized selector would then be holding a reference to the
         // instance, preventing the instance from being garbage collected, and that would be bad
         const withExtras = { ...props }
-        if (withRef) withExtras.ref = this.setWrappedInstance
+        //if (withRef) withExtras.ref = this.setWrappedInstance
         if (renderCountProp) withExtras[renderCountProp] = this.renderCount++
-        if (this.propsMode && this.subscription) withExtras[subscriptionKey] = this.subscription
+
         return withExtras
       }
+*/
+
+      renderInner(providerValue) {
+          const {storeState, store} = providerValue;
+
+          return (
+            <ConnectInner
+              key={this.version}
+              storeState={storeState}
+              store={store}
+              wrapperProps={this.props}
+            />
+          );
+      }
 
       render() {
-        if (this.state.error) {
-          throw this.state.error
-        } else {
-          return createElement(WrappedComponent, this.addExtraProps(this.state.props))
-        }
+          return (
+              <ReactReduxContext.Consumer>
+                  {this.renderInner}
+              </ReactReduxContext.Consumer>
+          )
       }
     }
 
     Connect.WrappedComponent = WrappedComponent
     Connect.displayName = displayName
-    Connect.childContextTypes = childContextTypes
-    Connect.contextTypes = contextTypes
-    Connect.propTypes = contextTypes
-    Connect.getDerivedStateFromProps = getDerivedStateFromProps
 
+    // TODO We're losing the ability to add a store as a prop. Not sure there's anything we can do about that.
+
```

Comment:
I still think we can manage this, but by allowing passing a separate context consumer to connect. This way, users can use a custom context provider and just hook into it with their specific components. In other words, we provide `createProviderContext` and allow:

```js
const { SpecialProvider, SpecialConsumer } = createProviderContext()
function MyProvider = createProvider(SpecialProvider)
const ConnectedNormal = connect(...)(Thing)
const ConnectedOther = connect(mSP, mDP, merge, { contextConsumer: SpecialConsumer })(Thing)

function Something() {
  return (
    <Provider store={store}>
      <Provider context={SpecialProvider} store={store2}>
        <ConnectedNormal>
          <ConnectedOther />
        </ConnectedNormal>
      </Provider>
    </Provider>
  )
}
```

This way, we get the full advantage of React's context, but have flexibility.